runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1306 @@
+/*
+ * Copyright 2023 Robert Bosch GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * \file cloe/data_broker.cpp
+ * \see  cloe/data_broker.hpp
+ *
+ * Classes:
+ *   DataBroker:
+ *     A central registry for (type erased) signals.
+ *   
+ *   Signal:
+ *     A type-erased abstraction of a signal.
+ *  
+ *   Container<T>
+ *     An optional container storing the value of a signal and abstracting the interaction with Signal.
+ *   
+ * Background:
+ *   Real world simulations utilize manifolds of signals (variables) and coresponding types. 
+ *   Still the very basic problem around a single signal (variable) boils down to CRUD
+ *   - declaring a signal (Create)
+ *   - reading its value / receiving value-changed events (Read)
+ *   - writing its value / triggering value-changed events (Update)
+ *   - NOT NEEDED: undeclaring a signal (Delete)
+ *
+ * Concept:
+ *   Abstract CRU(D) operations by registering signals in a uniform way by their name.
+ */
+
+#pragma once
+#ifndef CLOE_DATA_BROKER_HPP_
+#define CLOE_DATA_BROKER_HPP_
+
+#include <any>
+#include <cassert>
+#include <functional>
+#include <map>
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace cloe {
+
+namespace databroker {
+
+template <typename TValue>
+using on_value_change_event_std_function_t = std::function<void(const TValue*)>;
+template <typename TValue>
+using on_value_change_event_function_t = void (*)(const TValue*);
+template <typename TClass, typename TValue>
+using on_value_change_event_member_function_t = void (TClass::*)(const TValue*);
+
+}  // namespace databroker
+
+/**
+ * Predicate which determines whether a type is wanted/unwanted as signal datatype
+ */
+template <typename TValue>
+constexpr bool is_wanted_type =

```

Comment:
C++17 features: Done.

Philosophy-A: Code shall describe what it tries to achieve. Here: Mask wanted & unwanted types.
Philosophy-B: Code shall describe what it implements. Here: Having an unqualified_base_type.

I.e. for the moment I left the name unchanged.

Reply:
Agree, it's a matter of taste and philosophy :)

My perspective on this: I am probably biased by the STL way naming type trait functions. If you have a look at what the [type_traits header](https://en.cppreference.com/w/cpp/header/type_traits) has to offer, then all of the functions are named so that they describe properties of a type. (Even more so in the case of concepts, see [here](https://en.cppreference.com/w/cpp/language/constraints) and [here](https://en.cppreference.com/w/cpp/header/concepts))

Of course this is not directly comparable, as the STL wants to be general-purpose. In any case it was my feeling, that type trait functions are usually rather named based on properties / categories of the allowed types instead of intent.

Reply:
Since this is implementation i would probably also prefer is_unqualified, but is_compatible is also preferable over is_wanted, since wants is what a user has.

Reply:
Concept-refinement:
template <typename T>
constexpr bool is_incompatible_type_v = std::is_void_v<T> || std::is_reference_v<T>;

Reply:
I guess you'd have to check for pointer types too right? `std::is_reference_v<double*> ?: false`

Reply:
Conceptual:
If we design towards bare compatibility: Pointers are supported.
If we design towards desirability: is_incompatible_type_v got a naming issue.